src/aiida/manage/configuration/settings.py

```diff
@@ -66,45 +66,71 @@ def create_instance_directories() -> None:
         os.umask(umask)
 
 
-def set_configuration_directory(aiida_config_folder: pathlib.Path | None = None) -> None:
-    """Determine location of configuration directory, set related global variables and create instance directories.
+def get_configuration_directory():
+    """Return the path of the configuration directory.
 
-    The location of the configuration folder will be determined and optionally created following these heuristics:
+    The location of the configuration directory is defined following these heuristics:
 
-        * If an explicit path is provided by `aiida_config_folder`, that will be set as the configuration folder.
-        * Otherwise, if the `AIIDA_PATH` variable is set, all the paths will be checked to see if they contain a
-          configuration folder. The first one to be encountered will be set as `AIIDA_CONFIG_FOLDER`. If none of them
+        * If the ``AIIDA_PATH`` variable is set, all the paths will be checked to see if they contain a
+          configuration folder. The first one to be encountered will be set as ``AIIDA_CONFIG_FOLDER``. If none of them
           contain one, a configuration folder will be created in the last path considered.
-        * If the `AIIDA_PATH` variable is not set the `DEFAULT_AIIDA_PATH` value will be used as base path and if it
-          does not yet contain a configuration folder, one will be created.
+        * If ``AIIDA_PATH`` is not set, the current working directory is checked for an existing configuration directory
+          moving up the file hierarchy until the first is encountered or the root directory is hit.
+        * If an existing directory is still not found, the ``DEFAULT_AIIDA_PATH`` is used.
 
-    In principle then, a configuration folder should always be found or automatically created.
+    :returns: The path of the configuration directory.
     """
-    global AIIDA_CONFIG_FOLDER  # noqa: PLW0603
-    global DAEMON_DIR  # noqa: PLW0603
-    global DAEMON_LOG_DIR  # noqa: PLW0603
-    global ACCESS_CONTROL_DIR  # noqa: PLW0603
-
-    if aiida_config_folder is not None:
-        AIIDA_CONFIG_FOLDER = aiida_config_folder
+    dirpath_config = None
 
-    elif environment_variable := os.environ.get(DEFAULT_AIIDA_PATH_VARIABLE):
-        # Loop over all the paths in the `AIIDA_PATH` variable to see if any of them contain a configuration folder
+    if environment_variable := os.environ.get(DEFAULT_AIIDA_PATH_VARIABLE):
```

Comment:
The logic here is still relatively complex, perhaps it would be worth to split the two if branches into two small functions (`get_configuration_directory_from_envvar` and `get_configuration_directory_from_cwd` perhaps) , to make `get_configuration_directory` less busy.